libexec/rtld-elf/riscv/reloc.c

```diff
@@ -83,6 +89,54 @@ init_pltgot(Obj_Entry *obj)
 	}
 }
 
+#ifdef __CHERI_PURE_CAPABILITY__
+/*
+ * Plain RISC-V can rely on PC-relative addressing early in rtld startup.
+ * However, pure capability code requires capabilities from the captable for
+ * function calls, and so we must perform early self-relocation before calling
+ * the general _rtld C entry point.
+ *
+ * TODO: For now we use __cap_relocs. Instead we should use normal ELF
+ *       relocations that are all assumed to be relative capability
+ *       relocations, ditching the ad-hoc __cap_relocs format and using
+ *       CBuildCap.
+ */
+void _rtld_relocate_nonplt_self(Elf_Dyn *dynp, Elf_Auxinfo *aux);
+
+void
+_rtld_relocate_nonplt_self(Elf_Dyn *dynp, Elf_Auxinfo *aux)
+{
+	caddr_t relocbase = NULL;
+	const struct capreloc *caprelocs = NULL, *caprelocslim;
+	Elf_Addr caprelocssz = 0;
+	void *pcc;
+
+	for (; aux->a_type != AT_NULL; aux++) {
+		if (aux->a_type == AT_BASE) {
+			relocbase = aux->a_un.a_ptr;
+			break;
+		}
+	}
+
+	for (; dynp->d_tag != DT_NULL; dynp++) {
+		switch (dynp->d_tag) {
+		case DT_RISCV_CHERI___CAPRELOCS:
```

Comment:
Will this work if we have more than 2 entries or do we need -fno-jump-tables then? I can't remember whether we have jump tables right now.

Reply:
Jump tables are currently disabled completely upstream. But we also have `CFLAGS.reloc.c+=-fno-jump-tables` in `libexec/rtld-cheri-elf/Makefile` that's not guarded by being MIPS, so we should be fine.